### ISSUE TITLE
lftp does not follow http redirect (302) in mirror mode

### ISSUE THREAD

**Issue by thozza** [lavv17/lftp#105]
I was not able to find any existing or closed issue related to this behavior.

When you run e.g.:
```
lftp -c '; mirror -c -P -v -v -v -I *.rpm -X "/headers/" -X "/repodata/" -X "*.src.rpm" -X "/SRPMS/" -X "*-debuginfo-*.rpm" -X "/debug/" http://packages.vmware.com/tools/esx/4.1u3/rhel6/x86_64/ /tmp/'
```
lftp ends with
```
mirror: File moved: 302 Moved Temporarily (/tools/esx/4.1u3/rhel6/x86_64 -> /tools/esx/4.1u3/rhel6/x86_64/index.html)
```

I would expect it to follow the redirection.

We have a proposed patch from our engineer
```
diff -up lftp-4.0.9/src/MirrorJob.cc.mirror302 lftp-4.0.9/src/MirrorJob.cc
--- lftp-4.0.9/src/MirrorJob.cc.mirror302	2009-12-21 17:31:53.000000000 +0100
+++ lftp-4.0.9/src/MirrorJob.cc	2013-03-28 14:51:16.391224874 +0100
@@ -538,23 +538,25 @@ void MirrorJob::HandleChdir(FileAccessRe
 	 // cd to another url.
 	 const char *loc_c=session->GetNewLocation();
 	 int max_redirections=ResMgr::Query("xfer:max-redirections",0);
-	 if(loc_c && max_redirections>0 && last_char(loc_c)=='/')
+	 if(loc_c && max_redirections>0)
 	 {
 	    if(++redirections>max_redirections)
 	       goto cd_err_normal;
 	    eprintf(_("%s: received redirection to `%s'\n"),"mirror",loc_c);
 
 	    char *loc=alloca_strdup(loc_c);
-	    session->Close(); // loc_c is no longer valid.
-
 	    ParsedURL u(loc,true);
 
 	    if(!u.proto)
 	    {
-	       url::decode_string(loc);
-	       session->Chdir(loc);
+	       bool is_file=(last_char(loc)!='/');
+	       FileAccess::Path new_cwd(session->GetNewCwd());
+	       new_cwd.Change(0,is_file,loc);
+	       session->PathVerify(new_cwd);
+	       session->Roll();
 	       return;
 	    }
+	    session->Close(); // loc_c is no longer valid.
 	    session=FA::New(&u);
 	    session->Chdir(u.path);
 	    return;
```

but it has some possible downsides:

1. the value of 'Location' field, that is mandatory field in the header of 302 return code, is specified as a absoluteURI in the HTTP RFC document. This specification doesn't solve distinguishing between file and directory.
2. the logic of lftp mirror command uses recursion for downloading the sub-directories.

**Comment by thozza** [comment]
I found out that there is an issue if there are directories in the path we are mirroring.

When running the command noted in the first comment I get:
```
mirror: received redirection to `/tools/esx/4.1u3/rhel6/x86_64/index.html'
mirror: received redirection to `/tools/esx/4.1u3/rhel6/x86_64/index.html/repodata/index.html'
mirror: Access failed: 404 Not Found (/tools/esx/4.1u3/rhel6/x86_64/index.html/repodata/index.html)
```
I think that if the last portion in the path is a file it should be stripped off before accessing some nested directories. However I'm still not that familiar with lftp source so have no patch yet.

Any ideas?

**Comment by lavv17** [comment]
please test b06344f

**Comment by thozza** [comment]
Subsequent redirections now work fine. Thank you for the quick fix.

**Comment by thozza** [comment]
Hi.

By further testing we discovered, that lftp has issues with redirection if the Location, to which the server redirects the client is a full URL. This can be tested for example if you setup Apache with
```
RedirectMatch /source/$ /target/index.html
```
and create the /var/www/html/target/index.html file

```
$ ./lftp -d -c 'mirror -c http://localhost/source/ /tmp'
---- Resolving host address...
---- 1 address found: 127.0.0.1
---- Connecting to localhost (127.0.0.1) port 80
---- Sending request...
---> HEAD /source/ HTTP/1.1
---> Host: localhost
---> User-Agent: lftp/4.6.1
---> Accept: */*
---> Connection: keep-alive
---> 
<--- HTTP/1.1 302 Found
<--- Date: Tue, 31 Mar 2015 13:13:46 GMT
<--- Server: Apache/2.4.10 (Fedora)
<--- Location: http://localhost/target/index.html
<--- Keep-Alive: timeout=5, max=100
<--- Connection: Keep-Alive
<--- Content-Type: text/html; charset=iso-8859-1
<--- 
mirror: received redirection to `http://localhost/target/index.html'
---- Sending request...
---> HEAD /target/index.html/ HTTP/1.1
---> Host: localhost
---> User-Agent: lftp/4.6.1
---> Accept: */*
---> Connection: keep-alive
---> 
<--- HTTP/1.1 404 Not Found
<--- Date: Tue, 31 Mar 2015 13:13:46 GMT
<--- Server: Apache/2.4.10 (Fedora)
<--- Keep-Alive: timeout=5, max=99
<--- Connection: Keep-Alive
<--- Content-Type: text/html; charset=iso-8859-1
<--- 
mirror: Access failed: 404 Not Found (/target/index.html)
---- Closing HTTP connection
```

The problem is that lftp thinks that the index.html is a file. From what I investigated, this is due to the new ParsedURL contains the proto. This causes session->Chdir() to be called which then calls Change() on new_cwd with new_is_file argument set to false.

I have a patch ready, will create a pull request. Please review if my change makes sense.

**Comment by lavv17** [comment]
fixed by 0f7273a52b4158c7193e44b128a9e54db74c7979
